src/composition/supergraph/config/federation.rs

```diff
@@ -0,0 +1,160 @@
+    #[derive(Clone, Debug)]
+    pub struct FromSubgraphs;
+}
+
+/// Error that occurs when the user-selected `FederationVersion` is within Federation 1 boundaries,
+/// but the subgraphs use the `@link` directive, which requires Federation 2
+#[derive(thiserror::Error, Debug, Getters)]
+#[error(
+    "The 'federation_version' specified ({}) is invalid. The following subgraphs contain '@link' directives, which are only valid in Federation 2: {}",
+    specified_federation_version,
+    subgraph_names.join(", ")
+)]
+pub struct FederationVersionMismatch {
+    /// The user specified federation version
+    specified_federation_version: FederationVersion,
+    /// The subgraph names that have requested Federation 2 features
+    subgraph_names: Vec<String>,
+}
+
+/// This is the harness for resolving a FederationVersion
+#[derive(Clone, Debug)]
+pub struct FederationVersionResolver<State: Clone> {
+    state: PhantomData<State>,
+    federation_version: Option<FederationVersion>,
+}
+
+impl Default for FederationVersionResolver<state::FromSupergraphConfig> {
+    fn default() -> Self {
+        FederationVersionResolver {
+            state: PhantomData::<state::FromSupergraphConfig>,
+            federation_version: None,
+        }
+    }
+}
+
+/// Represents a `FederationVersionResolver` that has been initiated from user input, if any
+/// and is ready to take into account a supergraph config file, resolve immediately, or proceed
+/// to a later stage
+pub type FederationVersionResolverFromSupergraphConfig =
+    FederationVersionResolver<state::FromSupergraphConfig>;
+
+impl FederationVersionResolver<state::FromSupergraphConfig> {
+    /// Creates a new `FederationVersionResolver` from a [`FederationVersion`]
+    pub fn new(
+        federation_version: FederationVersion,
+    ) -> FederationVersionResolver<state::FromSupergraphConfig> {
+        FederationVersionResolver {
+            federation_version: Some(federation_version),
+            state: PhantomData::<state::FromSupergraphConfig>,
+        }
+    }
+
+    /// Produces a new `FederationVersionResolver` that takes into account the [`FederationVersion`]
+    /// from a [`SupergraphConfig`] (if it has one)
+    pub fn from_supergraph_config(
+        self,
+        supergraph_config: &SupergraphConfig,
+    ) -> FederationVersionResolver<state::FromSubgraphs> {
+        let federation_version = self
+            .federation_version
+            .or(supergraph_config.get_federation_version());
+        FederationVersionResolver {
+            state: PhantomData::<state::FromSubgraphs>,
+            federation_version,
+        }
+    }
+
+    /// Skips [`SupergraphConfig`] resolution, presumably because there is none
+    pub fn skip_supergraph_resolution(self) -> FederationVersionResolver<state::FromSubgraphs> {
+        FederationVersionResolver {
+            state: PhantomData::<state::FromSubgraphs>,
+            federation_version: self.federation_version,
+        }
+    }
```

Comment:
Is it better/is there a way to merge this together into a single step, so we take an Option<SupergraphConfig>?

Reply:
I merged these, which I think is more consistent with what we do elsewhere.